### PR TITLE
Create materialized `latest_metrics` table

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?
 
Create a materialized metrics table (`latest_metrics`) that contains the summary value for each unique `(metric_key, run)` tuple. When a run is fetched via the `GetRun` API, this summary value should be used when populating the run's `data` field.

The addition of this table will remove the need for a complex query that fetches the most recent value for a particular `(metric_key, run)` tuple based on `step`, `timestamp`, and `value`, thus providing performance improvements.

This PR also introduces eager loading for run attributes when appropriate (e.g., during `search_runs` and `get_runs` invocations); this helps avoid repeated database accesses that occur when model object attributes are accessed under a lazy load paradigm.
 
## How is this patch tested?
 
This PR has been tested with a combination of unit tests (see new migration test) and manual tests. I have manually verified that the latest metrics migration functions against MySQL and PostGres databases. I have also confirmed that the optimizations introduced by this PR dramatically improve UI load times. For example, loading the UI with a *remote* SQLStore containing 400 runs with 10 metrics each used to take upwards of 20 seconds; it now takes between 1 and 2 seconds.
 
## Release Notes
 
The performance of the SQLAlchemy store's `Get Runs` and `Search Runs` APIs has been improved significantly, resulting in better UI performance.

### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [ ] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [X] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [ ] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/sqlalchemy_store.py [hunk 1]
```diff
@@ -426,6 +426,8 @@ def log_metric(self, run_id, metric):
             self._get_or_create(model=SqlMetric, run_uuid=run_id, key=metric.key,
                                 value=value, timestamp=metric.timestamp, step=metric.step,
                                 session=session, is_nan=is_nan)
+            # TODO(czumar): Update the value in the `latest_metrics` table. This change cannot be merged
```
Comment (dbczumar) [inline mlflow/store/sqlalchemy_store.py]: This update logic needs to be introduced before this change can be merged

**Review by smurching — COMMENTED** [review]
Looks good! Just had some small comments / questions

File: mlflow/server/handlers.py [hunk 2]
```diff
@@ -159,9 +159,6 @@ def _get_experiment():
     response_message = GetExperiment.Response()
     experiment = _get_store().get_experiment(request_message.experiment_id).to_proto()
     response_message.experiment.MergeFrom(experiment)
-    run_info_entities = _get_store().list_run_infos(request_message.experiment_id,
```
Comment (mparkhe) [inline mlflow/server/handlers.py]: Does this account for change in expected outcome? Would there be users that depend on runs returned with experiment?

File: mlflow/store/db_migrations/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+* ``466860`` total metric entries
+* ``186`` runs
+* An average of ``125`` entries per unique metric
+
+#### Recovering from a failed migration
+If the **create\_latest\_metrics\_table** migration fails, simply delete the ``latest_metrics``
+table from your Tracking database as follows:
+
+```sql
+DROP TABLE latest_metrics;
```
Comment (mparkhe) [inline mlflow/store/db_migrations/README.md]: `USE <database_name>;`

File: mlflow/store/db_migrations/versions/89d4b8295536_create_latest_metrics_table.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        "**IMPORTANT**: This migration creates a `latest_metrics` table and populates it with the"
+        " latest metric entry for each unique (run_id, metric_key) tuple. Latest metric entries are"
+        " computed based on step, timestamp, and value. This migration may take a long time for"
+        " databases containing a large number of metric entries. Please refer to {readme_link} for"
+        " information about this migration, including how to estimate migration size and how to"
+        " restore your database to its original state if the migration is unsuccessful. If you"
+        " encounter failures while executing this migration, please file a GitHub issue at"
+        " {issues_link}.".format(
+            readme_link=(
+                "https://github.com/mlflow/mlflow/blob/master/mlflow/store/db_migrations/README.md"
```
Comment (mparkhe) [inline mlflow/store/db_migrations/versions/89d4b8295536_create_latest_metrics_table.py]: :+1:

File: mlflow/store/sqlalchemy_store.py [hunk 5]
```diff
@@ -83,7 +83,8 @@ def __init__(self, db_uri, default_artifact_root):
             SqlMetric.__tablename__,
             SqlParam.__tablename__,
             SqlTag.__tablename__,
-            SqlExperimentTag.__tablename__
+            SqlExperimentTag.__tablename__,
+            SqlLatestMetric.__tablename__,
```
Comment (mparkhe) [inline mlflow/store/sqlalchemy_store.py]: :+1: on that trailing `,`

**Review by mparkhe — COMMENTED** [review]
Looks good. Just one comment about runs list with `get_experiment`

File: mlflow/store/db_migrations/README.md (hunk 3 above)
Comment (dbczumar) [reply mlflow/store/db_migrations/README.md]: `USE <database_name>` is a MySQL-specific instruction. For example, users select databases in Postgres via the `\c` instruction. For this reason, I'd rather not include the `USE` instruction.

File: mlflow/server/handlers.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/server/handlers.py]: As discussed offline, this field has been deprecated as of MLflow 1.2 and is not consistently populated across backend stores. The Search API is now the recommended mechanism for fetching runs. For these reasons, @mparkhe and I have agreed that depopulating here makes sense.

**Pull request merged**